compiler/fuzzer/src/coverage.rs

```diff
@@ -1,81 +1,82 @@
+use bitvec::prelude::*;
 use candy_vm::fiber::InstructionPointer;
 use itertools::Itertools;
 use std::{
     fmt,
     ops::{Add, Range},
 };
 
-pub struct Coverage {
-    covered: Vec<bool>,
-}
+pub struct Coverage(BitVec);
+
 impl Coverage {
     pub fn none() -> Self {
-        Self { covered: vec![] }
+        Self(bitvec![])
     }
 
     pub fn add(&mut self, ip: InstructionPointer) {
-        if *ip >= self.covered.len() {
-            self.covered
-                .extend([false].repeat(*ip - self.covered.len() + 1));
+        if *ip >= self.0.len() {
+            self.0.extend([false].repeat(*ip - self.0.len() + 1));
         }
-        self.covered[*ip] = true;
+        self.0.set(*ip, true);
     }
 
     pub fn is_covered(&self, ip: InstructionPointer) -> bool {
-        self.covered.get(*ip).copied().unwrap_or_default()
+        *self.0.get(*ip).unwrap()
     }
 
     pub fn improvement_on(&self, other: &Coverage) -> usize {
-        self.covered
+        self.0
             .iter()
-            .copied()
-            .zip(other.covered.iter().copied())
-            .filter(|(a, b)| *a && !b)
+            .zip(other.0.iter())
+            .filter(|(a, b)| **a && !**b)
             .count()
     }
 
-    pub fn relative_coverage_of_range(&self, range: Range<InstructionPointer>) -> f64 {
-        assert!(!range.is_empty());
-        let len = *range.end - *range.start;
-        range.filter(|ip| self.is_covered(*ip)).count() as f64 / len as f64
+    fn bitslice_in_range(&self, range: Range<InstructionPointer>) -> &BitSlice {
+        &self.0[*range.start..*range.end]
+    }
+    pub fn in_range(&self, range: Range<InstructionPointer>) -> Coverage {
+        Self(self.bitslice_in_range(range).to_bitvec())
```

Comment:
Instead of cloning the data, we could store a `Cow<'a, BitVec>` internally. However, this function is a bit dangerous since the returned Coverage doesn't work with the original `InstructionPointer`s. Maybe even add a separate struct `RangeCoverage` or `FunctionCoverage` that offers fewer methods and only stores a `&'a BitSlice`?